core/src/main/kotlin/org/eln2/data/MultiMap.kt

```diff
@@ -0,0 +1,231 @@
+package org.eln2.data
+
+/**
+ * A helper class for representing a mapping as a Map.Entry.
+ */
+data class MultiMapEntry<K, V>(override val key: K, override val value: V): Map.Entry<K, V>
+
+/**
+ * A MultiMap: a data structure that maps one key (K) to zero or more values (V).
+ *
+ * This is in contrast to the usual Map, which maps one key to at most one value.
+ *
+ */
+interface MultiMap<K, V> {
+	/**
+	 * Tests whether a key [k] is in the MultiMap.
+	 *
+	 * This is true if k is associated with at least one value.
+	 */
+	operator fun contains(k: K): Boolean = get(k).isNotEmpty()
+
+	/**
+	 * Get the set of all values associated with key [k].
+	 */
+	operator fun get(k: K): Set<V>
+
+	/**
+	 * Get a single, arbitrary value associated with key [k].
+	 *
+	 * If that association is empty, returns null.
+	 */
+	fun one(k: K): V? {
+		val iter = get(k).iterator()
+		return if(iter.hasNext()) iter.next() else null
+	}
+
+
+
+	/**
+	 * An iterator over every Entry<K, Set<V>>, of size keySize.
+	 */
+	val keyMapping: Iterable<Map.Entry<K, Set<V>>>
+
+	/**
+	 * The number of keys, value sets, and keyMapping entries in this MultiMap.
+	 *
+	 * A naive, O(n) implementation might be keyMapping.count(), but you are encouraged to implement a faster one if at all possible. Implementors should document their complexity.
+	 */
+	val keyMappingSize: Int
+
+	/**
+	 * The set of all keys.
+	 */
+	val keys: Set<K> get() = keyMapping.map {(k, _) -> k }.toSet()
+
+	/**
+	 * The set of all sets of values.
+	 * 
+	 * This will have the same size as keySize.
+	 */
+	val valueSets: Collection<Set<V>> get() = keyMapping.map {(_, vs) -> vs}
+
+	/**
+	 * The unique values in this MultiMap.
+	 *
+	 * This is often at least O(n * UNION), where UNION is the cost of set union.
+	 */
+	val uniqueValues: Set<V>
+
+	/**
+	 * The number of unique values in this MultiMap.
+	 */
+	val uniqueValuesSize: Int get() = uniqueValues.size
+
+	/**
+	 * An iterator over every Entry<K, V>, where K may be repeated for each value.
+	 */
+	val entries: Iterable<Map.Entry<K, V>>
+
+	/**
+	 * The number of pairings in this MultiMap.
+	 *
+	 * A naive, O(n) implementation might be entries.count(), but you are encouraged to implement a faster one if at all possible. Implementors should document their complexity.
+	 */
+	val entriesSize: Int
+}
+
+/**
+ * An implementor of [MultiMap] using the Kotlin stdlib.
+ *
+ * This class is defined here using a Map<K, Set<V>> idiom which is compatible with the Kotlin stdlib. Other, more performant implementations exist, and can be swapped in here if needed.
+ */
+class SetMapMultiMap<K, V>(iter: Iterator<Pair<K, V>>): MultiMap<K, V> {
+	/**
+	 * Internally implements the MultiMap interface as a Map<K, Set<V>>.
+	 */
+	val map: Map<K, Set<V>> = with(mutableMapOf<K, MutableSet<V>>()) {
+		iter.forEach { (k, v) -> getOrPut(k, { mutableSetOf() }).add(v) }
+		println("SMMM.map: $this")
+		entries.associate { (k, v) -> Pair(k, v.toSet()) }
```

Comment:
This could easily be extremely inefficient.

If the class exists as a comparison for tests/documentation/fuzz comparisons, that's fine, but in that case it should be annotated with @TestOnly. If it's currently in use, then... well, I suppose that's fine for now; just add a "TODO: Add performance tests".

Reply:
I've discovered it is much a similar problem as what occurs with Rust, and one of the motivations for their `MaybeInit<T>` union (that replaces a bunch of unsafe raw pointer scariness and `mem::uninitialized()`): even though the data structure is supposed to be immutable, _initializing it_ somehow tends to cause mutations (if you don't dig too deeply into the details). I have no idea how to fix this specifically in Kotlin, so I suspect I'll end up having to go for the "TODO" option.